src/devices/src/virtio/net/iovec.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use crate::virtio::DescriptorChain;
+use std::io::IoSlice;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use vm_memory::{GuestMemory, GuestMemoryError, GuestMemoryMmap};
+
+#[derive(Debug)]
+pub enum Error {
+    /// We found a descriptor with unexpected access rights
+    DescriptorAccessMode,
+    /// An error happened with guest memory handling
+    GuestMemory(vm_memory::GuestMemoryError),
+}
+
+impl From<GuestMemoryError> for Error {
+    fn from(err: GuestMemoryError) -> Self {
+        Error::GuestMemory(err)
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub(crate) struct IoVec<'a> {
+    // container of the memory regions included in this IO vector
+    vecs: Vec<IoSlice<'a>>,
+    // IoVecs can contain memory regions that are either read-only
+    // or write-only
+    is_write_only: bool,
```

Comment:
I was actually thinking to drop this altogether, in the sense that in our case the DescriptorChain should always be read only, since we only use this at the TX path.

Reply:
Can't we also use this approach on the RX path ?

Reply:
There are some complications in the RX path, but I think we can use it there as well. In case we do, one solution would be to follow the Rust standard library and create a `mut` version of our type like [IoSliceMut](https://doc.rust-lang.org/std/io/struct.IoSliceMut.html).